### DISCUSSION TITLE
Move to React Server Components where ever appropriate

### DISCUSSION THREAD

**Discussion by deo002** [eclipse-sw360/sw360-frontend#184] [Ideas]
﻿**Proposal: Refactor of SW360 UI to Render Appropriate Components Server-Side**  

**Introduction**

NextJS now supports **React Server Components (abbreviated RSC)** using the new **app** directory. RSCs give us a huge advantage over the Client Site Rendered Apps and the Server-Side Rendered Apps. 

Client Side Rendered Applications are very interactive (i.e., we can use React state hooks and immediately update UI) and give access to Browser APIs. 

The Server Side Rendered Applications provide several benefits such as Server-Side Data Fetching (closer to data source so low latency and no dealing with complex state management with UseEffect), caching components on server, reduces client JavaScript bundle size, faster initial page load and many more.

RSC allows us to have the best of both worlds by letting us decide which part of our code will run where (server-side or client side). 

More on this:

- https://nextjs.org/docs/app/building-your-application/rendering/server-components
- https://nextjs.org/docs/app/building-your-application/rendering/client-components
- <https://www.mux.com/blog/what-are-react-server-components> (gives an in-depth analysis and comparison of SSR, SPA and RSC)

Here’s what NextJS says on when to use client components and server components:
![Screenshot from 2023-12-08 17-29-08](https://github.com/eclipse-sw360/sw360-frontend/assets/59907159/f4a2f9a8-e4f5-495e-8b99-98dd7ca46886)

**Current Flow and Drawbacks:**

Currently, most of our components are client side rendered. 

Drawbacks:

- Not utilizing the full potential and performance benefits NextJS provides us with RSC.
- Dealing with complex Lifecycle and State management hooks such as UseEffect, useSession and UseState.  
- Increased size of client-side JavaScript bundle due to large dependencies being client-side. 
- Relatively slow page loading times.
- Client-Side Data Fetching. NextJS recommends fetching data on server whenever possible. (Source: <https://nextjs.org/docs/app/building-your-application/data-fetching/patterns#fetching-data-on-the-server>)

**Proposed Flow:**

Render appropriate components server-side. For example, let’s take the case of Vulnerability Detail’s page:
![Screenshot from 2023-12-08 17-31-24](https://github.com/eclipse-sw360/sw360-frontend/assets/59907159/dec1cb89-12ca-4b82-abe0-9a928142b595)
![Picture3](https://github.com/eclipse-sw360/sw360-frontend/assets/59907159/f2b2205f-b404-4e07-afd7-5c05af72234f)

|**Code Demo for Server Side Rendered Vulnerability Detail Page: <https://github.com/siemens/sw360-frontend/tree/rsc_demo1/src/app/%5Blocale%5D/vulnerabilities/detail/%5Bid%5D>**|
| :- |

More on Client and Server composition patterns: <https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns>

**Comment by heliocastro** [comment]
@deo002 Please can you write the PDF in .md format here ?

**Reply by deo002** [reply]
Sure, will do it!

**Comment by hoangnt2** [comment]
@heliocastro @deo002 I have a comment that using **useEffect** in the client components is necessary, it will help us control which functions will run when the state changes, not all functions. We shouldn't try to remove useEffect in client components.

There is a problem with **useEffect** that it will run twice. This problem occurs due to **react strict mode** and it only occurs in **development** mode (in **production** mode **useEffect** run once only). We can disable it by adding `reactStrictMode: false` configuration to **next.config.js** ([ref](https://nextjs.org/docs/app/api-reference/next-config-js/reactStrictMode)). But this action is not recommended.

**Reply by hoangnt2** [reply]
Of course, moving from client to server components is necessary work.

**Reply by deo002** [reply]
Agreed! For client components making API calls, we should use UseEffect. To overcome the problem of checking session before every usage, we can use getSession method(https://next-auth.js.org/getting-started/client#getsession). 
In general, we can use useSession for parts that require session in UI. For API calls, we can use getSession. What do you say @hoangnt2 @heliocastro ?

**Reply by hoangnt2** [reply]
@deo002 that sounds great, use getSession in fetch function is a good idea

**Comment by heliocastro** [comment]
Nothing against useEffect. Just against use it together with Fetch process.
There's obvious things like component setups that will require useEffect, but that's it. Is client work, nor server side fetch.

For getSession inside fetch, i'm not totally buying it, as we will end up in a concurrency issues. Worth try though.

**Reply by deo002** [reply]
What are your views on moving all fetch processes server side?

**Reply by heliocastro** [reply]
This is something that i would like, but then would not direct possible i think.
UNLESS we not do fetch on the component but in data element used by component. This can work.

Table will be the exception because pagination, but let's. not worry on that one right now.

**Reply by hoangnt2** [reply]
How about using [SWR](https://swr.vercel.app/) for data fetching.

**Reply by deo002** [reply]
We can use SWR. NEXT docs suggest it for [client side fetching](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-client-with-third-party-libraries). But the recommended way by Next docs is [server side fetching](https://nextjs.org/docs/app/building-your-application/data-fetching/patterns#fetching-data-on-the-server) where ever possible. 
We would have to see which one to choose from these 2.

**Reply by hoangnt2** [reply]
Sometime we need to fetch data on client component. Fetching data on the server and passing it as a prop to the client component is not a good idea. Use SWR for fetching data in client component will be better.

**Comment by hoangnt2** [comment]
@deo002 @heliocastro, we can try [use hook](https://github.com/acdlite/rfcs/blob/first-class-promises/text/0000-first-class-support-for-promises.md#usepromise), then we don't need to fetch in useEffect anymore. Like this
```
"use client";

import { Post } from "@/lib/types";
import { Inter } from "next/font/google";
import { use } from "react";

const inter = Inter({ subsets: ["latin"] });

const getPosts = async (): Promise<Post[]> => {
  const data = await fetch("https://jsonplaceholder.typicode.com/posts");
  const posts = await data.json();

  return posts;
};

export default function ClientPosts() {
  const posts = use(getPosts());

  return (
    <div className={inter.className}>
      <h1>Posts</h1>
      <ul>
        {posts.map((post) => (
          <li key={post.id}>{post.title}</li>
        ))}
      </ul>
    </div>
  );
}

```
I found this from [this blog](https://livecode247.com/demystifying-data-fetching-in-nextjs-13)
